advisories/2024-09-25_cometbft-bls12381-double-hash.md

```diff
@@ -0,0 +1,146 @@
+# BLS12-381 Wrong Signature Generation 
+
+CometBFT's BLS12-381 signatures were generated from a further round of hashing of the message. This additional step ended up affecting the second-preimage resistance of the signature scheme, because different objects yielded the same signature under the same key.
+
+## Description
+
+CometBFT's support for BLS signatures over the BLS12-381 curve is a set of functions wrapping over `blst`'s Go bindings. The signature method of type `PrivKey`, [here](https://github.com/cometbft/cometbft/blob/276996ad958475b69727be2c57d4d0d818849a55/crypto/bls12381/key_bls12381.go#L110),
+adds a conditional branch to discern whether the message to sign, a byte slice, is longer than 32 bytes; in such case, the slice is hashed with `sha256`, and `blst`'s `Sign` method (of type `blst.P2Affine`) is called on the message digest, instead of the original message.
+
+Not only is this addition superflous but it alters the strength of the hash-to-curve construction of the underlying `blst` library.
+
+The `blst` library strictly follows [RFC 9380, "Hashing to curve"](https://datatracker.ietf.org/doc/html/rfc9380) when it comes to hashing arbitrary strings to one of the two subgroups G1 or G2 of the BLS12-381 curve. This operation requires the input string to be first hashed to the underlying field, before mapping the resulting field element to the given group (in the present case, G2).  This is better seen by inspecting `blst`'s `Sign` method [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/bindings/go/blst.go#L503). In the absence of any optional arguments, as it is always the case in CometBFT (because it invokes the latter function always without optional arguments), `useHash` is always `true` as per `parseOpts` (and `augSingle` is `nil`). Therefore, the function `HashToG2` is always the one to be invoked in our case. `HashToG2` will then invoke the C function `blst_hash_to_g2`, as seen [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/bindings/go/blst.go#L2747C4-L2747C19). The latter function ends up calling `hash_to_field`, as it can be seen [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/src/map_to_g2.c#L388-L401).
```

Comment:
```suggestion
The `blst` library strictly follows [RFC 9380, "Hashing to curve"](https://datatracker.ietf.org/doc/html/rfc9380) when it comes to hashing arbitrary strings to one of the two subgroups G1 or G2 of the BLS12-381 curve. This operation requires the input string to be first hashed to the underlying field, before mapping the resulting field element to the given group (in the present case, G2).  This is better seen by inspecting `blst`'s `Sign` method [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/bindings/go/blst.go#L503). In the absence of any optional arguments, as is always the case in CometBFT (because it always invokes the latter function without optional arguments), `useHash` is always `true` as per `parseOpts` (and `augSingle` is `nil`). Therefore, the function `HashToG2` is always the one to be invoked in our case. `HashToG2` will then invoke the C function `blst_hash_to_g2`, as seen [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/bindings/go/blst.go#L2747C4-L2747C19). The latter function ends up calling `hash_to_field`, as it can be seen [here](https://github.com/supranational/blst/blob/cf754001ddd10c30c366a2d6337e2a1a82bd6acf/src/map_to_g2.c#L388-L401).
```